hyperspy/_signals/signal2d.py

```diff
@@ -605,3 +608,433 @@ def add_ramp(self, ramp_x, ramp_y, offset=0):
+            peaks.append(tuple(p_new))
+    peaks = np.array([np.array(p) for p in set(peaks)])
+    return clean_peaks(peaks)
+
+
+def find_peaks_stat(z, standard_deviation=1.):
+    """
+    Method to locate positive peaks in an image based on statistical refinement
+    and difference with respect to mean intensity.
+    Parameters
+    ----------
+    z : ndarray
+        Array of image intensities.
+    Returns
+    -------
+    ndarray
+        (n_peaks, 2)
+        Array of peak coordinates.
+    Notes
+    -----
+    Implemented as described in the PhD thesis of Thomas White (2009) the
+    algorithm was developed by Gordon Ball during a summer project in
+    Cambridge.
+    This version by Ben Martineau (2016), with minor modifications to the
+    original where methods were ambiguous or unclear.
+    """
+    from scipy.ndimage.filters import generic_filter
+    from scipy.ndimage.filters import uniform_filter
+    from sklearn.cluster import DBSCAN
+
+    def normalize(image):
+        """Scales the image to intensities between 0 and 1."""
+        return image / np.max(image)
+
+    def _local_stat(image, radius, func):
+        """Calculates rolling method 'func' over a circular kernel."""
+        x, y = np.ogrid[-radius:radius + 1, -radius:radius + 1]
+        kernel = x ** 2 + y ** 2 <= radius ** 2
+        stat = generic_filter(image, func, footprint=kernel)
+        return stat
+
+    def local_mean(image, radius):
+        """Calculates rolling mean over a circular kernel."""
+        return _local_stat(image, radius, np.mean)
+
+    def local_std(image, radius):
+        """Calculates rolling standard deviation over a circular kernel."""
+        return _local_stat(image, radius, np.std)
+
+    def single_pixel_desensitize(image):
+        """Reduces single-pixel anomalies by nearest-neighbor smoothing."""
+        kernel = np.array([[0.5, 1, 0.5], [1, 1, 1], [0.5, 1, 0.5]])
+        smoothed_image = generic_filter(image, np.mean, footprint=kernel)
+        return smoothed_image
+
+    def stat_binarise(image):
+        """Peaks more than one standard deviation from the mean set to one."""
+        image_rolling_mean = local_mean(image, 10)
+        image_rolling_std = local_std(image, 10)
+        image = single_pixel_desensitize(image)
+        binarised_image = np.zeros(image.shape)
+        stat_mask = image > (
+            image_rolling_mean + standard_deviation * image_rolling_std)
+        binarised_image[stat_mask] = 1
+        return binarised_image
+
+    def smooth(image):
+        """Image convolved twice using a uniform 3x3 kernel."""
+        image = uniform_filter(image, size=3)
+        image = uniform_filter(image, size=3)
+        return image
+
+    def half_binarise(image):
+        """Image binarised about values of one-half intensity."""
+        binarised_image = np.zeros(image.shape)
+        binarised_image[image > 0.5] = 1
+        return binarised_image
```

Comment:
again, `np.where` is your friend here - can replace the function in 1 simple line..